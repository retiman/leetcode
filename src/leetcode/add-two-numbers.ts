export interface ListNode {
  value: number;
  next: ListNode | null;
}

export function node2list(node: ListNode): number[] {
  const xs: number[] = [];
  let current: ListNode | null = node;
  while (current !== null) {
    xs.push(current.value);
    current = current.next;
  }
  return xs;
}

export function number2list(x: number): number[] {
  const s = x.toString(10 /* radix */);

  const xs: number[] = [];
  for (let i = s.length - 1; i >= 0; i--) {
    const d = parseInt(s.charAt(i), 10 /* radix */);
    xs.push(d);
  }

  return xs;
}
