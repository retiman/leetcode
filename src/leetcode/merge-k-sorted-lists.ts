// This class definition comes from the problem itself and we cannot change it, or else our submission will not be
// accepted.
export class ListNode {
  val: number;

  next: ListNode | null;

  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

export function list2node(xs: number[] | null): ListNode | null {
  if (xs === null) {
    return null;
  }

  let head: ListNode | null = null;
  let current: ListNode | null = null;

  xs.forEach(x => {
    const node = new ListNode(x);
    if (current === null) {
      current = node;
      head = node;
    } else {
      current.next = node;
      current = current.next;
    }
  });

  return head;
}

export function node2list(node: ListNode | null): number[] {
  if (node === null) {
    return [];
  }

  const xs: number[] = [];
  let current: ListNode | null = node;
  while (current !== null) {
    xs.push(current.val);
    current = current.next;
  }

  return xs;
}
