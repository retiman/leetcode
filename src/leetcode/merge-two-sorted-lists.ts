// This class definition comes from the problem itself and we cannot change it, or else our submission will not be
// accepted.
export class ListNode {
  val: number;

  next: ListNode | null;

  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

export function list2node(xs: number[]) {
  let head: ListNode | null = null;
  let current: ListNode | null = null;

  xs.forEach(x => {
    const node = new ListNode(x);
    if (current === null) {
      current = node;
      head = node;
    } else {
      current.next = node;
      current = current.next;
    }
  });

  return head;
}
