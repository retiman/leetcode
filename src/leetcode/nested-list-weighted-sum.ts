// This class definition comes from the problem itself and we cannot change it, or else our submission will not be
// accepted.
export class NestedInteger {
  private value: number | null;

  private values: NestedInteger[];

  // If value is provided, then it holds a single integer
  // Otherwise it holds an empty nested list
  constructor(value?: number) {
    this.value = value ?? null;
    this.values = [];
  }

  // Return true if this NestedInteger holds a single integer, rather than a nested list.
  isInteger(): boolean {
    return this.value !== null;
  }

  // Return the single integer that this NestedInteger holds, if it holds a single integer
  // Return null if this NestedInteger holds a nested list
  getInteger(): number | null {
    return this.value;
  }

  // Set this NestedInteger to hold a single integer equal to value.
  setInteger(value: number) {
    this.value = value;
  }

  //  Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
  add(elem: NestedInteger) {
    this.value = null;
    this.values.push(elem);
  }

  //  Return the nested list that this NestedInteger holds,
  // or an empty list if this NestedInteger holds a single integer
  getList(): NestedInteger[] {
    return this.values;
  }
}
