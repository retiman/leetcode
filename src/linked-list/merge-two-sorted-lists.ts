// DIFFICULTY: Easy
//
// You are given the heads of two sorted linked lists list1 and list2.
//
// Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.
//
// Return the head of the merged linked list.
//
// See {@link https://leetcode.com/problems/merge-two-sorted-lists/}
import { ListNode } from './common/list-node';
export { mergeTwoLists };

// SOLUTION:
//
// A straightforward solution works.
function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  let head: ListNode | null = null;
  let tail: ListNode | null = null;
  let left = list1;
  let right = list2;

  while (left !== null || right !== null) {
    let node = null;

    if (left === null && right !== null) {
      node = right;
      right = right.next;
    }

    if (left !== null && right === null) {
      node = left;
      left = left.next;
    }

    if (left !== null && right !== null) {
      if (left.val <= right.val) {
        node = left;
        left = left.next;
      } else {
        node = right;
        right = right.next;
      }
    }

    if (tail === null) {
      tail = node;
      head = node;
    } else {
      tail.next = node;
      tail = tail.next;
    }
  }

  return head;
}
