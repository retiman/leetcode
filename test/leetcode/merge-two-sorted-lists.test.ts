// You are given the heads of two sorted linked lists list1 and list2.
//
// Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.
//
// Return the head of the merged linked list.
//
// See https://leetcode.com/problems/merge-two-sorted-lists/
import { list2node, ListNode } from '../../src/leetcode/merge-two-sorted-lists';

describe('merge two sorted lists', () => {
  function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    let head: ListNode | null = null;
    let current: ListNode | null = null;
    let left = list1;
    let right = list2;

    while (left !== null || right !== null) {
      const node = new ListNode();

      if (left === null && right !== null) {
        node.val = right.val;
        right = right.next;
      }

      if (left !== null && right === null) {
        node.val = left.val;
        left = left.next;
      }

      if (left !== null && right !== null) {
        if (left.val <= right.val) {
          node.val = left.val;
          left = left.next;
        } else {
          node.val = right.val;
          right = right.next;
        }
      }

      if (current === null) {
        current = node;
        head = node;
      } else {
        current.next = node;
        current = current.next;
      }
    }

    return head;
  }

  test('run', async () => {
    expect(mergeTwoLists(null, null)).toBeNull();
    expect(mergeTwoLists(new ListNode(1), null)).toStrictEqual(new ListNode(1));
    expect(mergeTwoLists(null, new ListNode(1))).toStrictEqual(new ListNode(1));
    expect(mergeTwoLists(new ListNode(1), new ListNode(2))).toMatchSnapshot();
    expect(mergeTwoLists(list2node([1, 2, 4]), list2node([1, 3, 4]))).toMatchSnapshot();
  });
});
