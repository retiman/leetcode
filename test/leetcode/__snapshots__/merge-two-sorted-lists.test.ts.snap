// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge two sorted lists run 1`] = `
ListNode {
  "next": ListNode {
    "next": null,
    "val": 2,
  },
  "val": 1,
}
`;

exports[`merge two sorted lists run 2`] = `
ListNode {
  "next": ListNode {
    "next": ListNode {
      "next": ListNode {
        "next": ListNode {
          "next": ListNode {
            "next": null,
            "val": 4,
          },
          "val": 4,
        },
        "val": 3,
      },
      "val": 2,
    },
    "val": 1,
  },
  "val": 1,
}
`;
